### DISCUSSION TITLE
Help transfer learning ECAPA-TDNN

### DISCUSSION THREAD

**Discussion by gabrielziegler3** [speechbrain/speechbrain#1648] [Q&A]
Hi there,

I have been using the speaker verification model ECAPA-TDNN and it works great for speaker verification. 
I want to take this model and apply it to **other speech classification** tasks like sentiment classification. I believe I should be able to load and freeze the embedding layers up to the last layer where the classifier is trained and then retrain a classifier model from scratch using the pretrained ECAPA as input. Does that sound right? 
I have been looking at the speechbrain tutorials like [this one](https://colab.research.google.com/drive/1LN7R3U3xneDgDRK2gC5MzGkLysCWxuC3?usp=sharing) and it wasn't so clear to me how I could train this new classifier model. Could you please enlighten me regarding the pretraining process?

I'd also appreciate it if there is a more detailed example for transfer learning. Once I learn the proper way to do it, I would be happy to contribute with more examples on the topic.

**Comment by mravanelli** [comment]
Hi Gabriel,
in the repository you can find other examples where we used ECAPA-TDNN for
other tasks. For instance, language-id (
https://github.com/speechbrain/speechbrain/tree/develop/recipes/VoxLingua107/lang_id/hparams)
and urban sound classification (
https://github.com/speechbrain/speechbrain/tree/develop/recipes/UrbanSound8k/SoundClassification/hparams
).
For studying the transfer learning properties of the model, the simplest
way is to pretrain the encoder and train the classifier from scratch. You
can use the pretrainer class, as shown here:
https://github.com/speechbrain/speechbrain/blob/develop/recipes/VoxCeleb/SpeakerRec/hparams/verification_ecapa.yaml#L77
.
I hope this could be useful for you.

Best,

Mirco

**Reply by modcs7** [reply]
@mravanelli Dear Mravanelli 
I follow the link below to prepare the ecapa.yaml file 
https://github.com/speechbrain/speechbrain/blob/develop/recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa_tdnn.yaml
but I got error when I used 
lr_annealing: !new:speechbrain.nnet.schedulers.CyclicLRScheduler
    base_lr: !ref <base_lr>
    max_lr: !ref <max_lr>
    step_size: !ref <step_size>
the error was there is no current_lr attribute in CyclicLRScheduler class
any suggestion 
thanks

**Reply by modcs7** [reply]
![123](https://github.com/speechbrain/speechbrain/assets/62318912/69984b31-a2ce-4993-ac9b-468df3ad34dc)

**Comment by gabrielziegler3** [comment]
Hi Mirco,
Thanks a lot for your resources and fast reply!
For this task, would it best to create a new recipe and base this new recipe on the UrbanSound8k recipe or it would be simpler to try to adapt the dataset to the UrbanSound structure and try to use the UrbanSound recipe? 
I have tried adapting the dataset but it doesn't look like a good path to go. I am not sure if I am going through the wrong steps or if the recipes are highly dependent on the task's dataset...

**Comment by mravanelli** [comment]
I think the best approach is to start from an existing recipe (e.g,
UrbanSound , CommonLanguage, Voxceleb-spk-id). If you plan to use your own
dataset, however, remember to write your own data preparation script. The
data preparation script has to create three data annotation files in JSON
or CSV formats (train.csv, valid.csv, test.csv) needed by train.py.  This
step is documented in the basic speechbrain tutorials that you find online
(see, for instance, speaker classification from scratch). Once you have
done the data preparation, you can try to run your scripts (remember to
adjust the number of output classes to your targets). Very likely  you have
to do some hyperparameter tuning to make things working well for your task.

**Comment by gabrielziegler3** [comment]
Sounds great, Mirco. Thanks again for your time! It is clearer for me now.
Happy holidays!

**Comment by gabrielziegler3** [comment]
Hi @mravanelli,
The training has been working well, but the classifier is struggling a bit due to the imbalanced nature of my custom dataset.
I searched but couldn't find how to deal with this in the proper manner with speechbrain. Should I be using a sampler like the [WeightedRandomSampler](https://github.com/speechbrain/speechbrain/blob/4aaabfda07b4d42e85f5d97184d375ef574eff6d/speechbrain/dataio/sampler.py#L110)? If so, how could I add this to the yaml file?

Not sure if my questions are already answered in the documentation, but I couldn't find the answers.
My classes' distribution is: 
- class 0: 393
- class 1: 71
- class 2: 36

I appreciate any tips to mitigate this issue (loss function, classifier, sampler, etc...)

**Comment by gabrielziegler3** [comment]
Update:

I have been using `ReproducibleWeightedRandomSampler` to weigh the classes to deal with the imbalanced classes.

```python
...
# NOTE Create custom weight sampler                                             
from collections import Counter                                                 
                                                                                
def get_class_weights(dataset):                                                 
    labels = [sample["sentiment_encoded"].item() for sample in dataset]         
    classes_counts = sorted(Counter(labels).items())                            
    classes_n_samples = torch.tensor([count[1] for count in classes_counts])    
    classes_weights = 1. / classes_n_samples                                    
    classes_weights = torch.tensor([classes_weights[t] for t in labels])        
    return classes_weights                                                      
                                                                                
train_class_weights = get_class_weights(datasets["train"])                      
                                       
from speechbrain.dataio.sampler import ReproducibleWeightedRandomSampler        
train_sampler = ReproducibleWeightedRandomSampler(                              
    weights=train_class_weights,                                                
    num_samples=len(train_class_weights),                                       
    replacement=True                                                            
)                                                                               
                                                                                
train_dataloader_options = hparams["dataloader_options"]                        
train_dataloader_options["sampler"] = train_sampler                             
                                                                                                         
sentiment_brain.fit(                                                            
    epoch_counter=sentiment_brain.hparams.epoch_counter,                        
    train_set=datasets["train"],                                                
    valid_set=datasets["valid"],                                                
    train_loader_kwargs=train_dataloader_options,                               
    valid_loader_kwargs=hparams["dataloader_options"],                          
    # progressbar=True,                                                         
)                                                                               
                                                                                
# Load the best checkpoint for evaluation                                       
test_stats = sentiment_brain.evaluate(                                          
    test_set=datasets["test"],                                                  
    min_key="error",                                                            
    progressbar=True,                                                           
    test_loader_kwargs=hparams["dataloader_options"],                           
)
...
```

It happens that I get the following error when speechbrain tries to run the validation step.

```python
speechbrain.utils.epoch_loop - Going into epoch 1
100%|█████████████████████████████████████████████████████████████████████| 125/125 [01:43<00:00,  1.21it/s, train_loss=6.79]
  0%|                                                                                                | 0/125 [00:00<?, ?it/s]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
~/SideProjects/speech-emotion-recogntion-ser2022/speechbrain/recipes/CORAA/train.py in <module>
    455     # print("DEBUG", hparams["dataloader_options"])
    456     sentiment_brain.fit(
--> 457         epoch_counter=sentiment_brain.hparams.epoch_counter,
    458         train_set=datasets["train"],
    459         valid_set=datasets["valid"],

~/PyEnv/DeepLearning/lib/python3.8/site-packages/speechbrain/core.py in fit(self, epoch_counter, train_set, valid_set, progressbar, train_loader_kwargs, valid_loader_kwargs)
   1069                 avg_valid_loss = 0.0
   1070                 with torch.no_grad():
-> 1071                     for batch in tqdm(
   1072                         valid_set, dynamic_ncols=True, disable=not enable
   1073                     ):

~/PyEnv/DeepLearning/lib/python3.8/site-packages/tqdm/std.py in __iter__(self)
   1178
   1179         try:
-> 1180             for obj in iterable:
   1181                 yield obj
   1182                 # Update and possibly print the progressbar.

~/PyEnv/DeepLearning/lib/python3.8/site-packages/torch/utils/data/dataloader.py in __next__(self)
    519             if self._sampler_iter is None:
    520                 self._reset()
--> 521             data = self._next_data()
    522             self._num_yielded += 1
    523             if self._dataset_kind == _DatasetKind.Iterable and \

~/PyEnv/DeepLearning/lib/python3.8/site-packages/torch/utils/data/dataloader.py in _next_data(self)
    559     def _next_data(self):
    560         index = self._next_index()  # may raise StopIteration
--> 561         data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
    562         if self._pin_memory:
    563             data = _utils.pin_memory.pin_memory(data)

~/PyEnv/DeepLearning/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py in fetch(self, possibly_batched_index)
     47     def fetch(self, possibly_batched_index):
     48         if self.auto_collation:
---> 49             data = [self.dataset[idx] for idx in possibly_batched_index]
     50         else:
     51             data = self.dataset[possibly_batched_index]

~/PyEnv/DeepLearning/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py in <listcomp>(.0)
     47     def fetch(self, possibly_batched_index):
     48         if self.auto_collation:
---> 49             data = [self.dataset[idx] for idx in possibly_batched_index]
     50         else:
     51             data = self.dataset[possibly_batched_index]

~/PyEnv/DeepLearning/lib/python3.8/site-packages/speechbrain/dataio/dataset.py in __getitem__(self, index)
    161
    162     def __getitem__(self, index):
--> 163         data_id = self.data_ids[index]
    164         data_point = self.data[data_id]
    165         return self.pipeline.compute_outputs({"id": data_id, **data_point})

IndexError: list index out of range
```

This code can be seen in this commit [61e43cb1](https://github.com/gabrielziegler3/speechbrain/blob/61e43cb1edf836b55510855b844ac9da56d57ce7/recipes/CORAA/train.py#L418).

I believe this is something I am doing wrong when passing the `ReproducibleWeightedRandomSampler`, but I am not sure what.
